File: example/typescript/SomeModule/Fixtures/Builders.ts
namespace SomeModule.Builder {
    export interface SomeClassDef {
        id?: string,
        amount?: number,
    }
    export function someClass(def?: SomeClassDef): SomeClass {
        return new SomeClass(
            new SomeId(def?.id ?? "someValue"),
            def?.amount ?? 0,
        )
    }

    export interface SomeClass2Def {
        id?: string,
        enabled?: boolean,
        names?: string[],
        ids?: string[],
    }
    export function someClass2(def?: SomeClass2Def): SomeClass2 {
        return new SomeClass2(
            new SomeId(def?.id ?? "someValue"),
            def?.enabled ?? false,
            def?.names ?? [],
            (def?.ids ?? []).map(it => new SomeId(it)),
        )
    }

    export interface SomeClass3Def {
        class2Object?: SomeClass2Def,
        class2List?: SomeClass2Def[],
    }
    export function someClass3(def?: SomeClass3Def): SomeClass3 {
        return new SomeClass3(
            someClass2(def?.class2Object ?? {}),
            (def?.class2List ?? []).map(it => someClass2(it)),
        )
    }

    export interface SomeClass4Def {
        otherId?: string,
        otherClass?: OtherModule.Builder.OtherClassDef,
        otherIdList?: string[],
        otherClassList?: OtherModule.Builder.OtherClassDef[],
    }
    export function someClass4(def?: SomeClass4Def): SomeClass4 {
        return new SomeClass4(
            new OtherId(def?.otherId ?? "someValue"),
            OtherModule.Builder.otherClass(def?.otherClass ?? {}),
            (def?.otherIdList ?? []).map(it => new OtherId(it)),
            (def?.otherClassList ?? []).map(it => OtherModule.Builder.otherClass(it)),
        )
    }
}

File: example/kotlin/src/testFixtures/kotlin/com/some/pkg/somemodule/Builders.kt
package com.some.pkg.somemodule

import com.some.pkg.othermodule.OtherClassDef
import com.some.pkg.othermodule.api.OtherId
import com.some.pkg.othermodule.otherClass
import com.some.pkg.somemodule.api.*

data class SomeClassDef(
    var id: String = "someValue",
    var amount: Int = 0,
)
fun someClass(init: SomeClassDef.() -> Unit = {}): SomeClass {
    val def = SomeClassDef().apply(init)
    return SomeClass(
        id = SomeId(def.id),
        amount = def.amount,
    )
}

data class SomeClass2Def(
    var id: String = "someValue",
    var enabled: Boolean = false,
    var names: List<String> = emptyList(),
    var ids: List<String> = emptyList(),
)
fun someClass2(init: SomeClass2Def.() -> Unit = {}): SomeClass2 {
    val def = SomeClass2Def().apply(init)
    return SomeClass2(
        id = SomeId(def.id),
        enabled = def.enabled,
        names = def.names,
        ids = def.ids.map { it -> SomeId(it) },
    )
}

data class SomeClass3Def(
    var class2Object: (SomeClass2Def.() -> Unit) = {},
    var class2List: List<(SomeClass2Def.() -> Unit)> = emptyList(),
    var someEnum: SomeEnum = SomeEnum.VALUE_A,
)
fun someClass3(init: SomeClass3Def.() -> Unit = {}): SomeClass3 {
    val def = SomeClass3Def().apply(init)
    return SomeClass3(
        class2Object = someClass2(def.class2Object),
        class2List = def.class2List.map { it -> someClass2(it) },
        someEnum = def.someEnum,
    )
}

data class SomeClass4Def(
    var otherId: String = "someValue",
    var otherClass: (OtherClassDef.() -> Unit) = {},
    var otherIdList: List<String> = emptyList(),
    var otherClassList: List<(OtherClassDef.() -> Unit)> = emptyList(),
)
fun someClass4(init: SomeClass4Def.() -> Unit = {}): SomeClass4 {
    val def = SomeClass4Def().apply(init)
    return SomeClass4(
        otherId = OtherId(def.otherId),
        otherClass = otherClass(def.otherClass),
        otherIdList = def.otherIdList.map { it -> OtherId(it) },
        otherClassList = def.otherClassList.map { it -> otherClass(it) },
    )
}